### ISSUE TITLE
Dependencies resolution with `--minimal-versions`

### ISSUE THREAD

**Issue by matklad** [rust-lang/cargo#5657]
Implementation PR: https://github.com/rust-lang/cargo/pull/5200

**Steps:**

- Document best-practices (https://github.com/rust-lang/cargo/issues/5656)

**Unresolved questions:**

- do we want to "impose" this feature on the ecosystem? Currently, everything seems to work fine due to eager dependency resolution. Adding `--minimal-versions` has costs: one-time ecosystem transition cost, cost to run CI job for minimal versions, cost to actually update minimal versions. There's anecdotal evidence that wrong minimal versions actually are a problem: https://www.reddit.com/r/rust/comments/8ob598/rust_minimum_versions_semver_is_a_lie/e027mtz/. 

- should we implement "--minimal-versions-for-me-but-not-my-dependencies" as well, to make the initial roll-out of this feature easier?

**Stabilization TODO:**

- change `-Z minimal-versions` to just `--minimal-versions` and add it alongside `--frozen/locked`,
- this will require community-wide effort to make the things work, so a special announcement should be prepared

**Comment by Eh2406** [comment]
I would like to cc some persons:
- @klausi you did the implementation and many of the prs to get CI using it, [including hear on cargo](https://github.com/rust-lang/cargo/pull/5275).
- @illicitonion you wrote up a [great blog post](https://blog.illicitonion.com/2018/06/rust-minimum-versions-semver-is-lie.html) on kicking the tyres on this feature.
- @matklad for your, [already linked](https://www.reddit.com/r/rust/comments/8ob598/rust_minimum_versions_semver_is_a_lie/e027mtz/), articulit descriptions of the downsides of "imposeing" this on the ecosystem.
- @alexcrichton as you have [encouraged some care](https://github.com/rust-lang/cargo/issues/5289#issuecomment-379513210) when it comes to [rolling](https://github.com/rust-lang/cargo/pull/5275#issuecomment-377965911) out this feature.
- @dtolnay as you have seen real world bugs caused by not having this.
- Who did I miss that holds strong views for/against this?

I'd like to see us come to a shared understanding of what we would recommend users do with this feature both at rollout and at steady state in the idealized future.
- A comically-pro-this-feature opinion thinks a crate is only valid if it can be built with `minimal-versions`, so we should check at `publish` time with no override, and we should `yank` all existing crates that don't meet this standard.
- A comically-against-this-feature opinion thinks that the meaning of `Cargo.toml` is what cargo dose with it, so if `">=0.0"` as a version requirement gets `cargo build` to work than what is the harm, So we should not stabilize this feature.

I think we are all in the middle, and probably closer than we think, we just need to articulate it. To start it off I will propose a position as a strawman, so you can point out where I am wrong.

All crates should try `minimal-versions`. If it can be made to work it should be added as a separate test in CI. As then things that depend on your crate will be able to use `minimal-versions` easily. `yank`ing is generally not called for, only to be used when it predates `rust 1.0.0` and it is a real problem for the ecosystem. If CI resources are tight, then `minimal-versions` can be combined with one of the other runs, like the minimal rust version or the beta run.

**Comment by klausi** [comment]
I'm currently not that interested in how strongly we should impose this option on the ecosystem. First we should make the minimal-versions option work properly. Not even cargo itself can be compiled with minimal-versions as explored in #5275.

Repeating the open points here:

1. `cargo publish` always needs to send the `links` key to the registry. This needs to be stabilized as default behaviour in stable cargo.
2. Mass update the crates.io registry to add the `links` key where it is missing right now.
3. Stabilize the `-Z minimal-versions` flag into `--minimal-versions` on Cargo stable

**Comment by Eh2406** [comment]
Well one small peace is done. At this time `cargo publish` from stable pushes the links key.

**Comment by klausi** [comment]
Oh cool, that is good news!

The inconsistent crates.io registry is still a problem. @alexcrichton said in #5289 that getting minimal-versions to work smoothly was not a priority then, so he didn't want to mass update the creates.io index. Is that still the assumption?

@SimonSapin argued that people will always use old cargo publish versions and by that make the crates.io registry inconsistent again. In order to prevent that we would have to implement server side rules on crates.io to prevent entries without `links` attributes. Which means breaking those old cargo publish versions.

For experimenting I'm maintaining a crates.io registry fork, that I update very irregularly. It has the `links` attribute fixed for some popular crates such as curl-sys.
```
CARGO_REGISTRY_INDEX=https://github.com/klausi/crates.io-index cargo build -Z minimal-versions
```

**Comment by Eh2406** [comment]
> Is that still the assumption?

That is one of the questions I think we need to address. :-) Hopefully, @alexcrichton and @matklad will have a chance soon to articulate what there current thoughts are.

What we decided to recommend **needs** to work and fairly smoothly. That means that if there are hard changes that need to be made to make it work then we need to be willing to do them. Correspondingly if we are not willing to make the big changes then we need to recommend something smaller.

Big changes may include some or all of:
- foundational and otherwize stable crates releasing new versions just to bump minimal dep.
- yanking versions too old to be worth making a new release for.
- manually updating the index to add the `links` attribute, and requiring it to be correct for all future uploads.
- other ideas?

Recommend something smaller may include some or all of:
- recommending a `--minimal-versions-for-me-but-not-my-grand-dependencies`
- endorsing a community maintained fork of the index that simulates the "Big changes"
- making it clear that this is an optional "nice to have" and that it is not worth doing unless your dependencies are already doing it.
- having some kind of log of things cargo discover do not work, that gets included into the index if they get tried again. (this would require a lot of design work.)
- other ideas?

**Comment by alexcrichton** [comment]
I think we'll definitely want to get this working with the main index, and I think we probably just want to keep going as-is, fixing up crates and publishing them as we discover mismatches. In that sense I think it might be good to do some more work to get some of the "base crates" working and then perhaps make a post on internals asking for testers so we can discover new crates and publish new versions

**Comment by illicitonion** [comment]
I got a non-trivial crate (https://github.com/pantsbuild/pants/tree/c8b42cb52eca9acbca98eaaf9599a47bc7b1f51e/src/rust/engine) compiling with `-Z minimal-versions`, and have started sending out a few PRs to the ecosystem.

My two big questions from this process are:
1. What version of the `log` crate should libraries depend on? Nothing before `0.3.4` builds with post-1.0 rust. `protoc` depends on `0.*` (https://github.com/stepancheg/rust-protobuf/blob/264debff3cd1cb26048925e90bec998941c2a328/protoc/Cargo.toml#L17), and many crates depend on `0.3`. Most things which depend on `0.3` work with `0.4` It would be great to have some published advice on this one, as it's a slightly weird dependency in the first place. `log` itself notes _some_ compatibility guidelines (https://docs.rs/log/0.4.2/log/#version-compatibility), but it would be nice to expand that to a firm suggestion ("You should always depend on the most recent version where possible" or "You should always depend on the oldest version which works for your library", or something else)
2. If we decide that `--minimal-versions` is something people should strive to support, how should people handle dependencies which aren't actively maintained, and which need updating? If a transitive dependency is the only problem with an otherwise functioning dependency which hasn't been touched in 2-3 years, and which doesn't respond to PRs, what _should_ people do? Fork the crate? Try to get the bad transitive dep yanked? Give up?

**Comment by alexcrichton** [comment]
@illicitonion nice!

Those are indeed good questions too :). You can somewhat force the process by having *some* crate have a higher version bound (aka requiring `0.3.10` of `log` synthetically) but that's not a great solution to either problem. I think for now the best advice we'd have is "try to send a PR to the crate and get a new version published" but that indeed reduces the usability of this feature :(

**Comment by Eh2406** [comment]
Actually I quite like the suggestion, 

> If you minimal-versions build is broken by log then add a dependency on 'log = "3.10"' to your Cargo.toml.
> If you can't do to your crate already depending on 'log = "4"' then depend on the helper crate 'logs-that-works-with-minimal-versions = "1"' witch is just a Cargo.toml that has 'log = "3.10"'.

I don't think "try to send a PR to the crate and get a new version published" reduces the usability, I think it is begging the question "If a transitive dependency is the only problem with an otherwise functioning dependency, and which doesn't respond to PRs, what should people do?".

**Comment by alexcrichton** [comment]
A snag I've now thought of as well: from time to time crates will break due to language/compiler changes, but we're generally pretty good about ensuring that the most recent version on crates.io always builds and  point releases are updated. This means, however, that lots of crates' CI will break when that Rust version is published, because not everyone will say they require the newer version of the crate.

**Comment by matklad** [comment]
@alexcrichton I think that depends on the CI setup. It seems to me that long-term the CI job with `--minimal-versions` should also use the minimal supported Rust version(MSRV): this is needed so that crates can bump minimal supported rust version in minor release, and it also saves one CI job as well. 

However, the day when your MSRV supports `--minimal-versions` is far away; in the meantime, the following guideline should work:

  * if your MSRV has `--minimal-version`, add a single CI job which both sets MSRV and `--minimal-versions`
  * otherwise, add two jobs:
    - MSRV **with a lockfile** (otherwise CI breaks when your deps bump their MSRV)
    - `--minimal-version` with rust 1.xy, where `1.xy` is the first rust release to support `--minimal-versions`

**Comment by Eh2406** [comment]
@matklad in the meantime can it be done in one CI job now with `cargo +nightly generate-lockfile --minimal-versions && cargo +MSRV test`?

**Comment by matklad** [comment]
Excellent idea @Eh2406!

**Comment by alexcrichton** [comment]
@matklad makes sense to me! And I like @Eh2406's idea as well, that should make an excellent suggestion for how to best use this flag

**Comment by Eh2406** [comment]
This is a list of foundational crates with versions that do not build on modern rust. Purging these from tomls is the "startup cost" of getting `minimal-versions` working. This is in a format that can be copied into a tomls to fix each dep.

- `winapi = "0.2.7"`
- `libc = "0.1.x" dont know x yet`
- `log = "0.3.4"`

I will keep this up to date as I find more, and one day make a `working-with-minimal-versions-hack` crate with this as it's starting toml.

**Comment by Eh2406** [comment]
So @dwijnand asked me to write up how do figure out what to do when a `minimal-versions` CI job fails.

So hear gose.
1. Reproduce locally. This involves checking out the branch and running what the CI job does. IntelliJ and RLS has a habit of updating a lock (without `-Z minimal-versions`) file on Cargo.toml change, So just to be safe I tend to close my editer when investigating.
2. Identify the problematic dependency. This is often easy to do by reading the path where the errors occurred. For example [this](https://ci.appveyor.com/project/rust-lang-libs/cargo/build/1.0.5235/job/3vleeax5uiabf8d4) hit an error in `C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\libc-0.1.1\rust/src/liblibc/lib.rs:79:46` So that is `libc-0.1.1` It is significantly harder if it is a build script that errored, as it can be any build-dep of that dep.
3. Determine how that ended up in your tree. I find [cargo tree ](https://github.com/sfackler/cargo-tree) extreamly helpful for this. So I do `cargo +nightly generate-lockfile -Z minimal-versions && cargo tree -p libc:0.1.1 -i` Cargo tree also updates the lock (without `-Z minimal-versions`) if it is not fully up-to-date, so I always run the two commands in one line. Also the format for specifying a dep uses a `:` instead of a `-` so watch out for that.
4. Find something to add to your toml to fix the dep. Best case is that the most recent version of the thing you actually depend on builds with `minimal-versions` requiring to that can solve a large swath of problems. For example [`curl = "0.4.13+"` ](https://github.com/rust-lang/cargo/pull/5757#issuecomment-407436703). Next best is that newer versions the thing you actually depend on no longer require the thing that causes the problem. This can solve the immediate problem, but can release new problems. For example [`git2 = "0.7.3"`](https://github.com/rust-lang/cargo/pull/5757#issuecomment-407239594) Next is to add a synthetic deps. I find that opening a tab for each compatible version of a deb on crates.io makes it fairly easy to binary search for when a dep got bumped.

**Comment by ehuss** [comment]
As discussed in #6636, dealing with breakage is difficult and can take a large amount of time to diagnose and fix, with questionable benefits.  Generally `minimal-versions` won't be useful unless everyone is using it (which IMHO is unlikely).  

@Eh2406 brought up in the team meeting today an alternate implementation that only enforced minimal versions on direct dependencies.  (Essentially forcing `=` requirements.)  To me, this sounds like it would be much easier to use.  It sounds like something that might be worth experimenting with.

**Comment by stjepang** [comment]
We inadvertently broke some crates by publishing a Crossbeam release without checking that it builds with minimal dependencies: https://github.com/crossbeam-rs/crossbeam/issues/312

I'd love to have a minimal-versions check in our CI, but it's impossible to build Crossbeam because the whole crates ecosystem is broken, unfortunately. There are two things we should do, IMO:

1. Add an option like `--minimal-versions-for-me-but-not-my-dependencies` as @matklad suggested. That would allow us to at least have *some* checks in CI rather than nothing.

2. Add an automatic check on `cargo publish`. We can issue a warning if the crate doesn't build with minimal dependencies, and maybe later in the future starting issuing errors.

**Comment by dekellum** [comment]
Somewhat related to the process outlined by @Eh2406 above...

As a prerelease exercise, I maintain and check in a Cargo.lock on a dedicated minimal-versions branch. The maintenance of that lock file isn't pretty, but I'm getting faster at it every time I do it, and I've avoided a few bugs with my own minimum versions thus far.   What seems to work best for me is keeping a companion script with a bunch off `cargo update -p <package> --precise <downgrade-version>` lines: dekellum/body-image@c16efc33  Note this includes minimal direct dependencies as well as some interesting, working, minimal-ish transitive dependencies. 

Then I not-so-Continuously Integration [test that branch](https://travis-ci.org/dekellum/body-image/builds/496648247).

Said another way: if you are lucky enough to produce a minimal-version `Cargo.lock` file that actually builds, definitely check-in somewhere in VCS, so you can continue to use it or compare it with future attempts!

**Comment by Eh2406** [comment]
Testing `--minimal-versions` for Cargo on windows was "temporarily" removed in #6748 do to the way [rand](https://github.com/rust-random/rand/issues/741) is broken.

**Comment by ehuss** [comment]
I was reading the [public private dependencies RFC](https://github.com/rust-lang/rfcs/blob/master/text/1977-public-private-dependencies.md) and I noticed it mentioned this feature:

> `cargo publish` will resolve dependencies to the *lowest* possible versions in order to check that the minimal version specified in `Cargo.toml` is correct.

Just leaving a note here about this because I wasn't aware about that until now, and pub/priv deps might make this issue more relevant.

I'm a little uncertain about issuing warnings if `cargo publish` fails. If a build fails, it would have to build twice, and for some crates it may take a very long time to build twice.  If cargo only did minimal versions for direct dependencies, I would be comfortable with just building once with minimal deps and make it a hard error if it fails.

**Comment by BurntSushi** [comment]
I had to [remove the minimal version check from `regex`'s CI tests](https://github.com/rust-lang/regex/commit/8f9ca96a844b07e6b378573ec6fbac6cf9fe5430). The `rand` crate continues to, for example, advertise support for `libc 0.2.0` even though `rand` cannot work with that version of `libc`. (There may be other incorrect dependency specifications in `rand`, I'm not sure.)

**Comment by ehuss** [comment]
@BurntSushi yea, I think this feature is dead as-is for now.  Someone needs to implement minimal-versions-for-me-but-not-my-dependencies and see how it goes from there.  I'm uncertain how difficult that will be, I haven't looked at it myself.

**Comment by vks** [comment]
@BurntSushi FWIW, the incorrect libc version in Rand was fixed. It's true that the Rand CI is not set up to test for the minimal versions, but I think pull requests correcting incorrectly specified minimal versions are accepted.

**Comment by BurntSushi** [comment]
@vks Sorry, but I'm not going act as rand's unofficial CI by submitting PRs whenever downstream detects that rand's minimal versions aren't correct. I've either just been removing the minimal version check on my end, or removing `rand` as a dependency where possible.

**Comment by dekellum** [comment]
FWIW: In my active projects, including a public crate with Cargo.lock below, I've avoided updates to rand for ~5 months now, while taking other updates, due to induced duplicates (rand-core 0.3.1 trick-shim + 0.4.0) and its perceived stability threat.

https://github.com/dekellum/body-image/commits/dev/Cargo.lock

Meanwhile, I've infrequently but successfully continued, semi-manual minimal version testing:

https://github.com/dekellum/body-image/compare/dev...minimal-versions-8

**Comment by josephlr** [comment]
We just ran into an interesting bug (or feature, not sure) around `-Z minimal-versions`.

`getrandom` (the OS-specific code for `rand`) attempts to have correctly specified minimum versions by running [a check in our CI](https://github.com/rust-random/getrandom/blob/6b4925a76df4fc81bb3cd95e9432b7734310d7c3/.travis.yml#L96-L98):
```
cargo generate-lockfile -Z minimal-versions
cargo test
```

However, even when this passes, we still get bugs like https://github.com/rust-random/getrandom/pull/112. The basic problem can be demonstrated in the following example:
 - Crate `A v0.2` unconditionally depends on crate `B v0.2.2`
 - Crate `A v0.2` has an optional dependency on crate `C v0.1`
   - Could also be a target-specific dependency, or dev dependency 
 - Crate `C v0.1` then depends on `B v0.2.6`.

Now let's say Crate `A` starts using features added in `B v0.2.5`. Running `cargo generate-lockfile -Z minimal-versions` will select `B v0.2.6` as the "minimal" version of Crate `B`, so the above checks will pass, despite Crate `A` now having an incorrect version specification.

Does anyone know of a way to fix this? Is it possible to build the minimal version dependency tree for just the mandatory dependencies, or with a specific set of Cargo features?

**Comment by dekellum** [comment]
It would seem that `cargo -Z minimal-versions generate-lockfile` should be aware of, and lock results conditional upon, what features are selected either by default, or via additional feature arguments to that command. You also demonstrate that the target can effect default features.  But as of nightly-2019-09-25, the `generate-lockfile` command does not accept feature nor target related flags.

Beyond the getrandom case but I suspect related: how does `-Z minimal-versions generate-lockfile`  interact with workspaces, e.g. if it is workspace and CWD package aware, and can yield different results if run from different package directories within a workspace tree?  Consider the case where two workspace packages had different minimums for the same dependency.

**Comment by mathstuf** [comment]
There might need to be additional cargo syntax here (or I'm missing how this can be done today).

  - crate A has a version spec on its dependency X of `>= 0.0.x, < 0.y`
  - crate A exposes X data types in its API, but the name of the crate is not necessarily visible
  - crate B depends on version A (any version, doesn't matter)
  - crate B uses the exposed X data type

However, B uses X APIs that A does not need (so A isn't going to bump its minimum version), but B needs to bump the minimum version A needs to expose for it to work. Adding a direct dependency on X is liable to get caught up in `cargo-udeps` checks since the crate is not directly used.

Is there some way for B to bump the minimum version of X it needs from A somehow without depending directly on X (without direct usage)?

The realworld case I hit this with is `yaml-rust` and `serde_yaml` (`linked-hash-map` is X here).

I don't *think* this should block, but defer to maintainers about how bad the problem is. This seems like a hairy problem and adding `Cargo.toml` stuff hasn't seemed to be the easiest thing to do.

**Comment by mathstuf** [comment]
Digging in more, `serde_yaml` *does* require a new enough `linked-hash-map` itself, but it still ends up with the really old API from `yaml-rust`'s exposure. I guess a PR to `yaml-rust` might fix this specific problem, but is that problem something to worry about overall?

**Comment by mathstuf** [comment]
Also seeing this with `serde` and `serde_derive` when using the `derive` feature for `serde` itself. `serde` just requires `serde_derive = "1.0"` which ends up at 1.0.0 and doesn't know about many of the attributes used. I'll stop posting more examples here (though I'll write down instances of this ), but that's a more serious one than the yaml crate at least.

**Comment by kornelski** [comment]
I think this feature is salvageable. It's a matter of yanking a few popular crates that have too lax dependencies. 

I've started a crusade against `"*"` deps:

<img width="450" alt="Screenshot 2019-12-07 at 16 19 29" src="https://user-images.githubusercontent.com/72159/70377546-add66680-190d-11ea-83cc-12f853258307.png">

**Comment by Nemo157** [comment]
> Someone needs to implement minimal-versions-for-me-but-not-my-dependencies and see how it goes from there.

I finally got round to figuring out how to do this on my project:

```sh
cargo update $(cargo metadata --all-features --format-version 1 | jq -r '. as $root | .resolve.nodes[] | select(.id == $root.resolve.root) | .deps[].pkg | . as $dep | $root.packages[] | select(.id == $dep) | "-p", "\(.name):\(.version)"') -Z minimal-versions
```

**Comment by Arnavion** [comment]
@Nemo157 I don't think that command is completely correct. With a crate that has a single dependency `chrono = "0.4"`, the `cargo metadata` command emits `-p chrono:0.4.13`. Running `cargo update -p chrono:0.4.13 -Z minimal-versions` generates a lockfile with chrono 0.4.0 and num 0.1.0 (chrono 0.4.0 depends on `num = { version = "0.1" }`) instead of chrono 0.4.0 and num 0.1.42

It seems to me what it needs to do is start with a non-`-Z minimal-versions` lockfile, then run `cargo update -p chrono --precise 0.4.0`, but I can't see if `cargo metadata` has a way to give us `chrono 0.4.0`

**Comment by Nemo157** [comment]
Ah yes, I forgot to mention that you should start with a normal maximal versions lockfile. It also will not correctly update newly added dependencies-of-dependencies to their maximal version, but that would require an iterative solution to perform externally to Cargo, and doesn’t come up in practice much.

**Comment by Arnavion** [comment]
Starting with a normal maximal versions lockfile doesn't change what I wrote.

```
$ cargo update $(cargo metadata --all-features --format-version 1 | jq -r '. as $root | .resolve.nodes[] | select(.id == $root.resolve.root) | .deps[].pkg | . as $dep | $root.packages[] | select(.id == $dep) | "-p", "\(.name):\(.version)"') -Z minimal-versions

    Updating crates.io index
    Removing autocfg v1.0.0
    Updating chrono v0.4.13 -> v0.4.0
      Adding num v0.1.0
    Removing num-integer v0.1.43
    Removing num-traits v0.2.12
```

**Comment by Arnavion** [comment]
Here's a version that implements the algorithm I specified in my first comment's edit:

```rust
cargo-minimal() {
	rm -f Cargo.lock
	cargo update -Z minimal-versions
	dep_names="$(cargo metadata --no-deps --format-version 1 | jq '.packages[].dependencies[] | "\(.name)"' -r)"
	declare -a dep_commands
	for dep_name in "$dep_names"; do
		dep_commands+="$(cargo metadata --format-version 1 | jq ".packages[] | select(.name == \"$dep_name\") | \"cargo update -p $dep_name --precise \(.version)\"" -r)"
	done

	rm Cargo.lock
	cargo update
	for dep_command in "${dep_commands[@]}"; do
		$dep_command
	done
}
```

Output:

```
    Updating crates.io index
    Updating crates.io index
    Updating crates.io index
    Updating chrono v0.4.13 -> v0.4.0
      Adding num v0.1.42
      Adding num-iter v0.1.41
```

The lockfile has the desired outcome of chrono 0.4.0 and num 0.1.42

**Comment by untitaker** [comment]
@xfix has recently opened issues across multiple crates to yank versions that do not compile on latest Rust. Others before him have asked crates to be yanked by other criteria.

I think this is an insufficient way to drive forward development and feasibility of `--minimal-versions`, as it leaves the entire ecossystem in a broken state for users of older Rust versions: older versions are yanked, and newer versions are unlikely to compile due to requiring new features. I would urge y'all to reconsider this strategy.

**Comment by kornelski** [comment]
@untitaker There's nothing to worry about yanking of old broken crates. It helps users of both new and old Rust versions.

First of all, the crates that need to be yanked are mainly from 2015, and don't work with Rust 1.x at all. They're not crates for what we call Rust, but crates for a previous experimental unstable language from before the Rust 1.0.0 release.

Even if we needed to yank some non-prehistoric crates that actually work with Rust 1.x, then their users wouldn't be affected much, because yanked crates can still be downloaded and used by users who have a lockfile.

What about users of very old Rust who don't have a lockfile? They need the `minimal-versions` feature! It's almost impossible to compile any Rust crate with older Rust and latest-version dependency resolution. It's an endless whack-a-mole of dependencies-of-dependencies-of-dependencies that bumped MSRV in a minor version. Crates with too-new MSRV bumped in minor version are even more common and more breaking than crates that break `minimal-versions`. So even today, with all its flaws, `minimal-versions` works *better* for users of old Rust compiler than normal dependency resolution.

**Comment by untitaker** [comment]
>First of all, the crates that need to be yanked are mainly from <=2015, and don't work with Rust 1.x at all. They're not crates for what we call Rust, but crates for a previous experimental unstable language that predated Rust 1.0.0 release.

That's not really what's happening in the most recently filed issues, where crates are considered for yanking if they don't work with latest Rust.

>What about users of very old Rust who don't have a lockfile? They need the minimal-versions feature!

Can they use it though, since all the crates are yanked now?

**Comment by kornelski** [comment]
Yanking doesn't delete crates, only hides them. Users who used yanked crates can continue to use them, because crate versions locked in lockfiles are usable indefinitely, even if they're yanked.

**Comment by untitaker** [comment]
Of course, but I'm talking about people who use very old Rust and don't have a lockfile. They are less likely to be able to use any crate successfully as it has been yanked. Even if they manually pick the last supported version and "pin" it in their toml, it won't work.

**Comment by kornelski** [comment]
Users of old Rust who don't have a lockfile are unable to use crates.io any more. At all.

**Comment by kornelski** [comment]
I've computed MSRV for every version of every Rust crate. I've compiled all crates with all Rust versions after 1.18 (that's the oldest version I checked, because older Rust versions don't have JSON message output, are slow, and there are almost no crates compatible with 1.18, and there aren't even many working with anything before 1.31).

The results bleak. Rust version older than 3 releases starts getting problematic. 10 releases back is barely usable if you carefully groom a lockfile. 1.31 is the oldest version that has any imaginable utility.

https://lib.rs/stats#rustc

**Comment by xfix** [comment]
> That's not really what's happening in the most recently filed issues, where crates are considered for yanking if they don't work with latest Rust.

- log 0.1.\*, 0.2.\* and 0.3.0 are pre-1.0 crates and don't compile on any 1.x release.
- getopts 0.1.0 to 0.2.16 is an interesting case. According to lib.rs the MSRV of 0.2.17 is 1.19~, however I was able to build it with Rust 1.18.0. Checking the actual MSRV is annoying because ancient Rust releases are stuck in an infinite loop when downloading crates.io-index - they can build programs once there is an actual Cargo.lock, but the problem is that Rust 1.0.0 isn't compatible with new lockfiles. In any case, yanking doesn't affect packages that are already in Cargo.lock.
- debug-unreachable 0.0.1 to 0.0.4 are pre-1.0.0 crates. In case of 0.0.1 to 0.0.2 they don't compile on any 1.x release, for 0.0.3 to 0.0.4 they manage to build, however the macro itself is broken.
- xml5ever 0.16.0 has the same MSRV as xml5ever 0.16.1.
- html5ever 0.25.0 has the same MSRV as html5ever 0.25.1.

**Comment by kornelski** [comment]
I've been [compiling crates with `-Z minimal-versions` for testing MSRV](https://lib.rs/stats#rustc), and once I've worked around it pulling the worst offenders:

* bitflags < 1.1.0
* cc < 1.0.43
* cmake < 0.1.42
* encoding < 0.2.33
* libc < 0.2.66
* log < 0.3.8
* mime < 0.2.6
* miniz-sys < 0.1.10
* pkg-config < 0.3.14
* rand < 0.3.23
* rustc-serialize < 0.3.24
* time < 0.1.39
* void < 1.0.2
* winapi < 0.2.8

then it works reasonably well. So I think it's mostly a matter of yanking these few broken pre-Rust-1.0 crates, and the option will be useful enough. 

The rest of breakage is mostly due to `serde` and `cc` adding new features without bumping semver-minor version, and users using `serde = "1"`, but that is good — it is actually finding real problems of specifying too-old versions or too-new features.

**Comment by epage** [comment]
I also added to my list in https://github.com/rust-lang/libs-team/issues/72:

> Stablize `#[stable]` and have rustc produce warnings when API items with a `#[stable]` are used that are newer than the oldest compatible version

which should help catch these problems earlier

> The versions above that I've used as a cut-off for too-old are newer than absolute minimum, but I still think it's mostly a matter of yanking broken pre-Rust-1.0 crates, and the option will be useful enough to stabilize.

I don't know about others but I still think this feature requires too much cooperation among the ecosystem as-is and that we should modify this to only control the version for direct dependencies.

> The rest of breakage is mostly due to serde and cc adding new features without bumping semver-minor version, and users using serde = "1", but that is good — it is actually finding real problems of specifying too-old versions or too-new features.

I can't see how bumping patch rather than minor has anything to do with this.  Regardless of which field is bumped, its a problem with whether the version requirement is bumped when new features are used.

**Comment by kornelski** [comment]
Warning about mismatch between `rust-version` and `#[stable]` is going to be useful, but it's tangential to `minimal-versions`. It's about mismatch between features and versions of crates, not features and versions of Rust.

I keep bringing up pre-Rust-1.0 crates, because they break builds for reasons unrelated to their upstream crates. Their users may have used their public API correctly. `minimal-versions` checks if graph edges of the dependency graph are correct, but the pre-1.0-Rust crates are leaf nodes that are broken themselves. It's a different kind of breakage.

For minimal versions, bumping semver-minor instead of semver-patch is as crucial as the difference between bumping semver-minor and semver-major for maximal versions. If people don't care about the difference between major and minor versions, then updates cause breakage. When people don't care about the difference between minor and patch, then downgrades cause breakage.

**Comment by epage** [comment]
> Warning about mismatch between rust-version and #[stable] is going to be useful, but it's tangential to minimal-versions. minimal-versions is about mismatch between features and versions of crates, not features and versions of Rust itself.

No, this is a separate item from `package.rust-version` and `#[stable]`.  The idea is that we'd stablize `#[stable]` for people to use in their crates.  Cargo can then look up what the minimal version is for the version requirement (independent of `--minimal-versions`) and pass that to rustc.  rustc can then warn if you are using an API item that is newer than `--minimal-versions`.

> For minimal versions, bumping semver-minor instead of semver-patch is as crucial as the difference between bumping semver-minor and semver-major for maximal versions. If people don't care about the difference between major and minor versions, then updates cause breakage. When people don't care about the difference between minor and patch, then downgrades cause breakage.

Except what does that have to do with `--minimal-versions`?

**Comment by kornelski** [comment]
[Here's a longer explanation](https://users.rust-lang.org/t/psa-please-specify-precise-dependency-versions-in-cargo-toml/71277) why `semver-minor` plays the same critical role as `semver-major` when downgrades are involved (and `minimal-versions` does downgrades).

**Comment by epage** [comment]
That has nothing to do with my question and you should know I am familiar with that post as it was a reply to me.   `minimal-versions` only does downgrades that are compatible with the version requirement.  If people specify only `x` or `x.y`, then yes, I can see there being problems.  If you specify `x.y.z`, then its all the same.  That is very different than the blanket statements you are making and the FUD that you keep spreading around these crates.

Side note: as I've said before on this topic, there are still problems with `x.y` when "properly following semver" because you can be dependent on bug fixes.  Being able to build or not is a small consolidation if your runtime behavior is still broken.

**Comment by kornelski** [comment]
I've initially posted a link to the thread as a whole, but that missed the exact point about semver-minor downgrades, and only then I've realized we've already had this discussion. I worry that discussion whether semver makes sense is out of scope of this thread. Should we take this discussion elsewhere or just agree to disagree?

**Comment by epage** [comment]
Yes, I'm trying to not re-hash that previous conversation and only focus on the part relevant to `--minimal-versions`.

**Comment by epage** [comment]
Now that we have `-Zdirect-minimal-versions`, I've been giving it some thought in terms of how we should stablize it.

Users to optimize for
- "The ecosystem": ensure high quality and following of common norms like semver
- Contributors: wants things to "just work"
- User running `cargo install`: wants latest features, bug fixes, and security patches
- Official builds: wants latest features, bug fixes, and security patches

The main possible workflows I can think of are:
1. `Cargo.lock` is either not committed or it contains state independent of `-Zdirect-minimal-versions`.  A CI job runs that enforces `-Zdirect-minimal-versions`
  - "The ecosystem" is getting a distributed crater run on every CI job, helping to catch problems earlier and putting the collective pain of breakages to bear down on maintainers to fix things
  - Users get the latest of everything via `cargo install`
  - Official builds are made from what is being tested and can include the latest of everything
  - But contributors won't find out that they are using too-new of functionality until they post their PR, see the job failure, and root cause that its because they either can't use that version or need to bump the version requirement
2. `Cargo.lock` always maintains `-Zdirect-minimal-versions`
  - But it takes longer for "The ecosystem" to upgrade and there isn't the collective pain of everyone feeling it at once
    - `[lib]`s have more of an incentive to use `-Zdirect-minimal-versions`, so most likely `[[bin]]`s will find these
    - But `[[bin]]`s are further away from the buggy dependency, making it more expensive to root cause
  - Contributors get immediate feedback if they do something that is incompatible
  - Users get the latest of everything via `cargo install` since the lockfile is ignored
  - Official builds need a separate lockfile that is tested and then used for official builds to get the latest of everything

Potential ways of exposing `-Zdirect-minimal-versions`, assuming Workflow 2 as its more complicated
- Command-line flag only affecting the current run
  - As the lockfile will be compatible, other runs of commands won't force it to re-calculate (e.g. `rust-analyzer` running behind the scenes)
  - But contributors are likely to forget to use `-Zdirect-minimal-versions` when making version requirement changes
- Expand `package.resolver` to a table with this being a setting
  - But `package.resolver` is permanent setting that affects `cargo install` (rust-lang/cargo#10112), implying this should be but that means users aren't getting the latest of everything when installing, making `cargo install` act like `cargo install --locked`
- We make `-Zdirect-minimal-versions` sticky, storing its use in the lockfile and defaulting it to what is in the lockfile
  - Affects local development
  - Affects `cargo install --locked` only in that the lockfile is used
  - Does not affect `cargo install`

**Comment by jonhoo** [comment]
One worry I have with having this be adjustable (if that's the intent) in the "sticky in the lockfile" world is that it'll be confusing for any user that moves from one crate to another. They expect version selection to work one way (and choose either old or new dependency versions), but it works the other way, possibly even silently, and pain ensues. I'm not entirely sure how to mitigate that danger.

It's also worth noting that (I believe) both Go and npm have version resolution that chooses minimal versions rather than maximum versions by default. That's not necessarily an argument for us doing the same, but I wonder if their rationale for that choice is documented anywhere.

**Comment by epage** [comment]
> It's also worth noting that (I believe) both Go and npm have version resolution that chooses minimal versions rather than maximum versions by default. That's not necessarily an argument for us doing the same, but I wonder if their rationale for that choice is documented anywhere.

The quintessential golang post is https://research.swtch.com/vgo-mvs

See also the HN discussion: https://news.ycombinator.com/item?id=16433425

**Comment by kornelski** [comment]
One option to consider is treating dependency requirement syntax differently, and when a full version is specified, treat it as minimum:

```toml
serde = "1" # means maximum version
serde = "1.1.123" # minimum version instead
serde = "^1.1.123" # means maximum version again
```

This is a departure from what Cargo does currently and would need an opt-in mechanism (perhaps edition?), but OTOH it is more similar to what npm does.

**Comment by epage** [comment]
Since others may be interested, here is a blog post explaining why this isn't default: https://web.archive.org/web/20190408003011/http://aturon.github.io/2018/07/25/cargo-version-selection/

**Comment by carols10cents** [comment]
I expanded all the comments on this issue and then did `CTRL-F` for "path" and didn't find this concern raised, so I apologize if I've missed discussion of this somewhere but it doesn't look like I have.

I recently had an issue with a project that uses `croaring`, and I upgraded `croaring`, but I neglected to upgrade `croaring-sys` as well.  My project's lockfile kept it on an older version of `croaring-sys`, which caused my project to fail to build. The version specified in `croaring` for `croaring-sys` indicated that the older version of `croaring-sys` should have been compatible, but that wasn't actually true. Longer explanation over here: https://github.com/RoaringBitmap/croaring-rs/pull/124

So I thought "this is a job for `-Zdirect-minimal-versions`! I'll send in a PR adding that job to CI, show that it would catch the problem I had, and then fix the problem!" only my plan didn't work because when building `croaring` in `croaring`'s project, the `path` dependency on `croaring-sys` gets used even if `-Zdirect-minimal-versions` is specified. 

So it seems like the `path` dependency is preventing Cargo from even considering a minimally compatible version that might be on crates.io. I think when `-Zdirect-minimal-versions` is specified, Cargo should consider both existing versions on crates.io that might satisfy the constraint _as well as_ the `path` dependency, in case the dependency hasn't been published yet.

I can work up a minimal reproduction of this problem in a repo if that would be helpful.

**Comment by Nemo157** [comment]
I was hoping that `cargo minimal-versions` would deal with `path` dependencies, but it just has an open PR adding the ability: https://github.com/taiki-e/cargo-minimal-versions/pull/4. I can confirm that using that branch and `cargo minimal-versions --detach-path-deps check` successfully finds the issue in `croaring`.

**Comment by taiki-e** [comment]
> it just has an open PR adding the ability: https://github.com/taiki-e/cargo-minimal-versions/pull/4.

FYI, `--detach-path-deps` has been released in cargo-minimal-versions 0.1.20.

**Comment by daxpedda** [comment]
I noticed that `minimal-versions` also pulls in yanked versions.
Is that a bug or intentional?

**Comment by Eh2406** [comment]
It should not. Pleas open a new issue with steps to reproduce.

**Comment by ctz** [comment]
Does `-Zdirect-minimal-versions` work as intended?

I see:

```
$ cargo minimal-versions --direct --ignore-private check
info: running `cargo update -Z direct-minimal-versions`
    Updating crates.io index
error: failed to select a version for `zeroize`.
    ... required by package `aws-lc-rs v1.6.0`
    ... which satisfies dependency `aws-lc-rs = "^1.6"` of package `rustls v0.23.0-alpha.0 (/home/runner/work/rustls/rustls/rustls)`
versions that meet the requirements `^1.7` are: 1.7.0

all possible versions conflict with previously selected packages.

  previously selected package `zeroize v1.6.0`
    ... which satisfies dependency `zeroize = "^1.6.0"` of package `rustls v0.23.0-alpha.0 (/home/runner/work/rustls/rustls/rustls)`

failed to select a version for `zeroize` which could resolve this conflict
```

To clarify: rustls has a dependency `zeroize = "1.6.0"`, and a dependency on `aws-lc-rs` which itself has a dependency `zeroize = "1.7"`.

I would expect this to be fine: `direct-minimal-versions` should minimize the `rustls` -> `zeroize` version to 1.6.0, and do the normal resolution for `rustls` -> `aws-lc-rs` -> `zeroize`. But instead, that error is seen.

This is `1.77.0-nightly (635124704 2024-01-27)`

**Comment by epage** [comment]
Direct-minimal-versions picks the minimum version for your dependencies (so 1.6). As a consequence, if it can't find a way for transitive dependencies on zeorize that are compatible with 1.6, it errors. It can't / shouldn't force aws-lc-rs to use 1.6 when it says it needs 1.7.

**Comment by ctz** [comment]
Apologies, what I was expecting here was to use =1.6 for my immediate zeroize dependency, and then ?>=1.7 for aws-lc-rs. Sounds like I have misunderstood what the "direct" part of `direct-minimal-versions` means?

**Comment by epage** [comment]
Cargo unifies versions both because they may be public and to reduce build times. We could evaluate allowing duplicates in the future after public dependencies are stabilized and widely used (currently blocked on bugs in the compiler with no one stepping up to fix) but we'd have to tread very carefully because of the build time impact. We could easily avoid unifying in places it doesn't make sense and slow things down a lot.

**Comment by mathstuf** [comment]
Diagnostics when duplication of dependencies occurs may be appreciated anyways. Those using `minimal-versions` may accept it in order to verify their own minimal version requirements accurately. Thanks for the details though.

**Comment by epage** [comment]
We discussed this in a recent Cargo team meeting.

Use cases
- **primary** testing version requirements
- **workaround** MSRV-testing without MSRV-aware resolver (#9930) as there is a correlation between low versions and low MSRVs
  - For packages without declared MSRVs, `-Zminimal-versions` still might better serve some people

Once we have #9930, in theory that leaves "testing version requirements".  An alternative solution to that is summed up as:

> drag manifest forward rather than dragging lockfile backwards

The idea is we try to transition the ecosystem over to #10498 by default so that manifests have your lockfile in it and has latest dependencies available.  At that point, there is no more minimal you can go and we don't need this.

This is blocked on
- #10498
- MSRV-aware resolver (#9930)

Potential problems with this
- Some users may want to hold back their version requirements as much as possible to give users more freedom in picking versions
  - As your dependency tree grows, the number of versions that can't be used together may increase.  Being forced to use the latest, even if its not needed, can block these kinds of workarounds
- As mentioned in https://github.com/rust-lang/cargo/issues/10498#issuecomment-2276604441, this would cause more churn in `Cargo.toml` that would cause more merge conflicts which some project may try to avoid
- As mentioned in https://github.com/rust-lang/cargo/issues/10498#issuecomment-2276604441, this would cause more friction for people who want latest, MSRV-incompatible dependencies during development but keep their version reqs MSRV-compatible.
- Lack of `package.rust-version` and working around that with `-Zminimal-versions`

**Comment by epage** [comment]
If we keep `-Zminimal-versions` in some form, we need to decide what that form is

The `-Zminimal-versions` of today is is easy to describe at a high level but that description will either not do what people want or, worse, people will think it does what they want (minimal version testing).   *Anything* (transitive or direct dep) can have a version higher than what you want, messing up minimal version testing.

`-Zdirect-minimal-versions` is harder to describe at a high level but it fits perfectly for minimal version testing.  However, its a pain because you need your version reqs to be the highest of your dep tree (it effectively pins to the lowest version)
- Maybe we could describe this with multiple fields in config?
- Maybe some judicious `cargo update` logic could make `-Zdirect-minimal-versions` less of a pain?

Maybe the way to expose `-Zminimal-versions` is similar to the idea of  #10498: help keep the lockfile and the manifest in-sync.  Taking this a step further, one thought I had is `cargo update` when `-Zminimal-versions` is enabled, we update go ahead and update dependencies but force `Cargo.toml` to be updated so the update lockfile content **is** the new minimal version.

**Comment by nyonson** [comment]
If I am working on a very conservative library which wants to ensure its small dependency tree (this might not be practical for large trees) does not have any minimal version issues, does it make sense to use both flags? A process like the following:

1. Run `rm -f Cargo.lock && cargo check -Z direct-minimal-versions`.
2. Run `rm -f Cargo.lock && cargo check -Z minimal-versions`.

The thinking is step one ensures the listed direct dependency versions are not quietly bumped up by transitive requirements. It fails fast and forces a maintainer to list the "real" minimal version of the dependency. Step two then creates the absolute minimum version of the whole dependency tree to test with, a more exhaustive test than the `direct-minimal-versions` versions which allow transitives versions to upgrade beyond their minimums.

Ideally, this would be one option instead of two assuming the pattern is appropriate.

**Comment by epage** [comment]
Just saw that pnpm has `lowest-direct` mode

See https://pnpm.io/settings#resolution-mode

**Comment by kornelski** [comment]
Golang only uses transitive minimal dependency resolution. Golang users are very happy with it, and praise it for making dependency updates reliable and predictable, without need for lockfiles.

I think there are couple of reasons why a similar strategy ends up having such different results in Golang and Cargo:

1. The minimal-versions resolution is the norm in Golang, so packages naturally test and correct for it. In Rust it's a rare edge case, so packages often allow too-low versions without realizing it. Stabilizing minimal-versions in Cargo could help, making it easier to test.

2. Duplicate semver-major versions are allowed only for packages that opt in to that (have a version in their URL, Golang calls it "Semantic Import Versioning"), and otherwise there's only one global version of a dependency. This means in practice it's usually easy to bump version of a transitive dependency.

I think the second point makes all the difference. In Cargo it's just as easy to fix too-low versions of transitive dependencies *within a single semver-major* version. Getting `serde 1.0.0` is not a big deal, because you can work around that by adding `serde = "1.0.300"` to your direct deps. However, the crates.io ecosystem has plenty of packages with multiple semver-major versions, and getting `log 0.3.0` can't be fixed by simply adding `log = "0.4.0"` to dependencies.

This makes me think that the real problem Cargo has with transitive minimal dependency resolution is not that it's transitive, but that users can only control semver-minor versions of transitive dependencies, and can't directly prevent wrong semver-major versions being picked via transitive dependencies.

This independence of semver-major versions also causes other annoyances, like unnecessary duplication of major versions when broad ranges are used, e.g. `itertools = "> 0.8, < 0.15"` won't unify with `itertools = "0.11"`.

So maybe the fragility of `-Z minimal-versions` could be mitigated by giving users a way to avoid duplicate/unwanted semver-major versions of dependencies anywhere in the dependency tree? `-Z direct-minimal-versions` works better because users easily control their direct deps, but controlling transitive deps could be made easier too! (and also help other issues users are having with transitive deps)

- For example, what if dependency resolution algorithm prioritized avoiding duplicate semver-major versions? (I mean this internally inside the resolver, without effect visible outside of which versions get picked when generating the lockfile).
	- run dependency resolution as far as possible without allowing duplicate semver-major versions, and relax that requirement only after it causes a dead-end (internally, still during the resolution phase, not during a build!) 
	- or run dependency resolution as normal. Then, if there are two major versions of the same crate, add a requirement that upgrades the oldest major version, see if the dependencies still resolve, keep the requirement if they do.

- or how about allowing users to specify required version range across major versions somehow? it could be `log = { version = "0.4", major-version = ">= 0.4" }`, or `log = { version = "0.4", globally-unique-version = true }`

- or add a `Cargo.toml` or `config.toml` section that lists unwanted crates (applying to transitive ones too), like a mini `cargo-deny`:
   ```toml
    [anti-dependencies] 
    log = "< 0.4.0"
    gcc = "*"
   ```

**Comment by reneleonhardt** [comment]
Great points!

>       * or run dependency resolution as normal. Then, if there are two major versions of the same crate, try force-upgrading the oldest major version and see if the dependencies still resolve.

But dependency versions are not everything in cargo, there are also rust versions 😉

I think another part of the problem is that crates don't have to declare their `rust-version`, can the dependency resolution even determine if the current toolchain supports a newer transitive dependency without compiling it if the crate's metadata doesn't contain their required MSRV?
Maybe a bisect like cargo-msrv would be needed in such cases, making resolution very slow...


I'm wondering if a "light-speed" cargo build would be possible, which would not be a build, only the fastest rustc compilation with the built-in "[profile.msrv]" but without linking and running, just to automatically determine the MSRV before publishing the crate if Cargo.toml doesn't declare rust-version.

**Comment by mathstuf** [comment]
I don't think you can assume that a build failure can only come from "the version doesn't work, let's try another". It could be any number of things, so I would be very hesitant about involving any build attempt results back into the dependency resolution logic. Kind of feels like Rust's type inference failures not rolling back and instead just hard-erroring. Basically, there is no SFINAE in Rust's type system; I really don't think we would want to adopt it for dependency resolution.

**Comment by kornelski** [comment]
@reneleonhardt I wasn't suggesting involving builds in this. My suggestions were for the internal dependency resolution algorithm only, which from user perspective still runs only once. The retrying approach is just an implementation detail. I imagine tweaking the resolver to prefer minimizing the number of redundant semver-major versions all at once while performing dependency resolution as normal would be difficult to implement. An optional fixup step can be added much easier, without having to add any new global optimization goal to the resolver.

Determining the right MSRV for crates is not easy, and [I know very well how many false positives there can be](https://gitlab.com/lib.rs/main/-/blob/683bbd1a70b85b47fb9898536b09982a4a8d308f/builder/src/parse.rs#L848). It's also hard to get all crate authors to specify it reliably. However, these are much bigger problems. `minimal-versions` is only one tool in the toolbox for dealing with all of it.

**Comment by reneleonhardt** [comment]
Thank you for clarifying your suggestions, I only wanted to point out that the resolver can't always be correct / compatible if only considering dependency versions, no matter what digit is being considered.

As an example, huggingface released breaking changes in hf-hub 0.4.2 without declaring a rust-version (MSRV was 1.81 in 0.4.1, then 1.82 in 0.4.2).
So, semver-minor isn't guaranteed to be a fence for undeclared breaking changes either 😄

**Comment by kornelski** [comment]
I agree that there are crates with missing or invalid `rust-version` and even cases where MSRV isn't monotonic. But I don't get what that is supposed to mean for `minimal-versions`. It obviously can't solve all the MSRV breakage 100% of the time, and we don't have any alternative that does that without running into other edge cases.

And note that `minimal-versions` isn't strictly for working around MSRV problems. It's useful when diagnosing semver breakage and other cross-crate incompatibilities, especially when bisecting or rebuilding older libraries that didn't have a lockfile (you can start with the lowest bound, and then selectively update deps one by one with `-p`).

It's also useful for ensuring dependency requirements have correct lower bound specified, so that they don't break when projects use a mix of non-latest versions for any other reason, e.g. projects that have to have all dependencies reviewed via `cargo-vet` tend to update very selectively, and generally lag a lot with updates, so when they do update deps, they need them to specify their requirements correctly instead of assuming every other dep is the latest. Larger projects may also avoid YOLOing `cargo update` of everything to avoid having duplicate semver-major deps, especially where too-new "public" deps cause breakage (like rustcrypto's `Digest` or various `HttpsConnector` impls). `minimal-versions` doesn't exactly solve it, but it's easier to correctly upgrade deps by hand starting from minimal versions, than to selectively downgrade transitive deps starting from all the latest ones.

**Comment by epage** [comment]
From @kornelski 

> Golang only uses transitive minimal dependency resolution. Golang users are very happy with it, and praise it for making dependency updates reliable and predictable, without need for lockfiles.

The way I read your post, it makes it sound like you are interested in shifting the Rust ecosystem over to MVS.  Considering the push back I got with committing lockfiles, I'm guessing that will be a hard sell and could actually get in the way of ever stabilizing `-Zminimal-versions`.

Also, have there been any studies done on upgrade speed in the Go ecosystem?

From https://research.swtch.com/vgo-mvs#upgrade_timing

> Given that minimal version selection takes the minimum allowed version of each dependency, it’s easy to think that this would lead to use of very old copies of packages, which in turn might lead to unnecessary bugs or security problems. In practice, however, I think the opposite will happen, because the minimum allowed version is the maximum of all the constraints, so the one lever of control made available to all modules in a build is the ability to force the use of a newer version of a dependency than would otherwise be used. I expect that users of minimal version selection will end up with programs that are almost as up-to-date as their friends using more aggressive systems like Cargo. 

This seems like a bold claim and it would be nice to see if its backed up with numbers.  I know in Rust, I rarely release new versions due to dependency upgrades and wouldn't appreciate the extra effort in the loop to do so to help encourage others to upgrade.  People have specifically opened issues against clap because it used to depend on `regex` and `regex` had that DoS issue.  They wanted clap to update its lockfile and I could close the PRs because the lockfile only affected clap and not clap users and clap wasn't affected.

From @kornelski 

> I think the second point makes all the difference. In Cargo it's just as easy to fix too-low versions of transitive dependencies within a single semver-major version. Getting serde 1.0.0 is not a big deal, because you can work around that by adding serde = "1.0.300" to your direct deps. However, the crates.io ecosystem has plenty of packages with multiple semver-major versions, and getting log 0.3.0 can't be fixed by simply adding log = "0.4.0" to dependencies.

I feel like I'm missing something as I'm not aware of a minimal versions problem with semver-major versions and the discussion around `log` doesn't sound related to `-Zminimal-versions`.

From @kornelski 

> For example, what if dependency resolution algorithm prioritized avoiding duplicate semver-major versions? (I mean this internally inside the resolver, without effect visible outside of which versions get picked when generating the lockfile).

From my understanding, adding constraints like this blows up the big-O of the resolver which is already not ideal in its performance, particularly in scaling.

From @kornelski 

> This independence of semver-major versions also causes other annoyances, like unnecessary duplication of major versions when broad ranges are used, e.g. itertools = "> 0.8, < 0.15" won't unify with itertools = "0.11".
>
> ...
>
> or how about allowing users to specify required version range across major versions somehow? it could be log = { version = "0.4", major-version = ">= 0.4" }, or log = { version = "0.4", globally-unique-version = true }

I'm a bit confused at this because you can do this today and I'm assuming you have done this yourself.  Granted, there are issues with it (e.g. #14582) and testing the combinations is difficult (`-Zminimal-versions` is insufficient).  I don't see how adding new flags would make a difference.  We definitely shouldn't have packages specifying unbounded version ranges.

For `globally-unique-version`, the issue for that is #5920.

From @kornelski 

> or add a Cargo.toml or config.toml section that lists unwanted crates (applying to transitive ones too), like a mini cargo-deny:

We will likely need something like this at some point, whether through #7193 or lists of some kind, its unclear.

**Comment by kornelski** [comment]
I'm not interested in moving the ecosystem to MVS. Updating to latest is the better default, and it's great that the crates.io ecosystem already managed to make it work pretty reliably. Golang's approach requires packages to release updates just to propagate transitive dependency updates, and this is a major downside that I wouldn't want to apply to the whole ecosystem. I brought up Golang not as something to emulate, only as a counter-example/contrast to the problems that `-Zminimal-versions` has in crates.io ecosystem.

I would like the crates.io ecosystem to have *more accurate* information about minimum versions required, because inaccurate information breaks or complicates some workflows. The problems aren't only limited to unusual workflows, and can affect even some basic operations. For example, `cargo add crate` can end up with a broken build if the lockfile already had the deps the crate needed, but in minor versions older than the new crate actually required. Similarly `cargo update -p crate` can update too few crates if the updated crate started depending on new APIs from newer dependencies without updating its `Cargo.toml` accordingly. Such breakage can be fixed with `cargo update`, but there are situations where a blanket `cargo update` is not desirable, such as use of `cargo-vet`/`cargo-crev` (you may need to update one thing and not have time to review everything else right now).

So `minimal-versions` isn't a goal in itself, but it's a useful tool. It works pretty well for uncovering too-low dependency requirements, which in turn helps crates work more reliably with `cargo add` and `cargo update -p` (and there are other use-cases I've mentioned previously).

>  the discussion around log doesn't sound related

It's an example of one of the problems users face when using `-Zminimal-versions`. Cargo supports having multiple different semver-major versions of deps, which is an incredibly important and useful feature. But it works much better with latest versions than minimal versions. 

Most crates you have across your dependency tree eventually update their deps, so with latest versions usually all the major versions end up aligning to the latest version, with some stragglers having a before-latest version in a worse case. But with minimal-versions there's no even bottom to align to. You get many different semver-major versions picked for different crates in the dependency tree. This alone isn't a problem yet — you get what you asked for with `minimal-versions`. However, you may need semver-major versions match across the dependency tree (for public deps, singletons), or prevent some very old crates from being used. Then it becomes a problem, because Cargo doesn't have easy ways of doing that. 

For example, latest-versions will easily give you latest `log 0.4` everywhere. But there are crates that upgraded from `log 0.3` to `log 0.4` in a minor/patch release, and downgrading them will pull in an undesirable `log 0.3`. You can't prevent that by having `log = "0.4"` as your direct dependency. You can't fix it by running `cargo update -p log`. You have to find every crate that may pull in `log 0.3`, and update it, and/or update the crate that pulled in that crate.

The difficulty of controlling duplicate/too-low semver-major versions comes up often when using `-Zminimal-versions`, but it's not exclusive to `-Zminimal-versions`. Even when using latest versions, users may still struggle to remove unnecessary duplicate crates or align "public" deps across crates that share them (e.g. different `http` crate versions). This is why I think it's not a flaw in `minimal-versions`, and suggested general-purpose solutions for controlling transitive deps versions.

**Comment by epage** [comment]
> I'm not interested in moving the ecosystem to MVS. U

Thanks for the clarification of your intent.  I agree its useful to learn from others and agree it would be helpful to have more accurate minimal versions. 

>  For example, cargo add crate can end up with a broken build if the lockfile already had the deps the crate needed, but in minor versions older than the new crate actually required.

For brainstorming purposes: two radical ideas
- (somehow) when adding a new dependency, we aggressively update all of its dependencies
- When the lockfile is changed, act as if `cargo generate-lockfile` is called, making the lockfile more of a consistent snapshot in time than a hodgepodge of different versions.  MSRV and various versions constraints would mean that there would still be some inconsistency.  People would still need to have the ability to override the resolve to specific versions to workaround undeclared MSRVs, bugs, etc.

> So minimal-versions isn't a goal in itself, but it's a useful tool. It works pretty well for uncovering too-low dependency requirements, which in turn helps crates work more reliably with cargo add and cargo update -p (and there are other use-cases I've mentioned previously).

An imperfect tool in that your workspace members and transitive dependencies can mask issues as well as ecosystem-wide cooperation.  direct-minimal helps with both but is more brittle.  Hmm, there is a gap with direct-minimal in terms of re-exports.  A way of resolving that is "direct" stays direct not just for the resolver roots dependencies but also their public dependencies, recursively.

> For example, latest-versions will easily give you latest log 0.4 everywhere. But there are crates that upgraded from log 0.3 to log 0.4 in a minor/patch release, and downgrading them will pull in an undesirable log 0.3. You can't prevent that by having log = "0.4" as your direct dependency. You can't fix it by running cargo update -p log. You have to find every crate that may pull in log 0.3, and update it, and/or update the crate that pulled in that crate.

That seems like `-Zminimal-versions` is working in that case if its to help with checking your lower bounds and its helping you find where your lower bounds are causing you to end up in a bad state.  Granted, there may be workflow improvements for identifying what the proper lower bound should be (or maybe just using latest is the answer).  As for reducing the chance of being in that bad state generally, it sounds like you also agree that that is a wider concern.

>  This is why I think it's not a flaw in minimal-versions, and suggested general-purpose solutions for controlling transitive deps versions.

So why is this being discussed here rather than one of those issues?

**Comment by briansmith** [comment]
> * When the lockfile is changed, act as if `cargo generate-lockfile` is called, making the lockfile more of a consistent snapshot in time than a hodgepodge of different versions.  MSRV and various versions constraints would mean that there would still be some inconsistency.  People would still need to have the ability to override the resolve to specific versions to workaround undeclared MSRVs, bugs, etc.

Many teams have a process where dependency updates need to be made individually using `cargo update -p`. (These teams would love a way to disable `cargo update` without specifying a list of packages.) It is important that no changes to the lockfile are made beyond what is requested.

Further, merge conflicts would become unmanageable.

**Comment by epage** [comment]
> Many teams have a process where dependency updates need to be made individually using cargo update -p.

This is the first I've heard of this and that workflow tends to have a lot of problems that I'm surprised aren't being brought up if it is "many teams".

Also, making an argument for or against something with an unnamed group without being able to interact to better understand their current workflow or explore what the overall workflow can be makes this feedback non-actionable.

**Comment by briansmith** [comment]
> > Many teams have a process where dependency updates need to be made individually using cargo update -p.
> 
> This is the first I've heard of this and that workflow tends to have a lot of problems that I'm surprised aren't being brought up if it is "many teams".

To clarify, "made individually" doesn't mean literally one-by-one. Usually indirect dependency updates are grouped into direct dependency updates. In other words, I am describing the default behavior of dependabot.
